### ISSUE TITLE
PoolManager tries to dequeue from empty Queue 

### ISSUE THREAD

**Issue by tbrown1979** [http4s/http4s#1920]
Issue we were seeing on production:
```scala
java.util.NoSuchElementException: queue empty
    at scala.collection.mutable.Queue.dequeue(Queue.scala:66)
    at org.http4s.client.PoolManager$$anonfun$borrow$1.go$1(PoolManager.scala:173)
    at org.http4s.client.PoolManager$$anonfun$borrow$1.apply(PoolManager.scala:187)
    at org.http4s.client.PoolManager$$anonfun$borrow$1.apply(PoolManager.scala:144)
    at io.chrisdavenport.scalaz.task.instances.TaskInstances$$anon$1$$anonfun$async$1.apply(TaskInstances.scala:28)
    at io.chrisdavenport.scalaz.task.instances.TaskInstances$$anon$1$$anonfun$async$1.apply(TaskInstances.scala:26)
```

#### Scenario (a more thorough explanation is below this): 

> We `createConnection` 10 times for the _same_ `RequestKey` so we're at the `MaxTotalConnections` then 1 connection is `release`ed. 
> 
> We release that 1 connection and we have nothing in the `WaitQueue` so we add to the `IdleQueue` for the `RequestKey`. This does not decrement the `curTotal`, which is still `10` and equal to the `MaxTotal`.
> 
> We then `createConnection` again, with a _different_ `RequestKey`.
> 
> We'll match on the `case None if curTotal == maxTotal =>` case. 
> 
> Then we will evict a random value from the idle queue. 
> 
> This will dequeue from the only idle queue we have available. 
> 
> Then we'll try to `decrConnection`. This _won't_ remove the key from the `idleQueue` because we have more allocated connections with the same `RequestKey`. 
> 
> Then you `borrow` again and you can end up dequeuing from the queue that has nothing in it?

#### We can distill this it a bit:

If the `idleQueue` contains a `RequestKey` mapping that only has _1 connection in the idleQueue_ **_and_** the `allocated` mapping has `> 1` connections for the **_same_** `RequestKey` then it is possible that we'll [dequeue](https://github.com/http4s/http4s/blob/master/client/src/main/scala/org/http4s/client/PoolManager.scala#L173) from that `idleQueue` mapping but we _won't_  [remove the key](https://github.com/http4s/http4s/blob/master/client/src/main/scala/org/http4s/client/PoolManager.scala#L65). If another `borrow` is called then it is _possible_ that we'll try to `"evict"/dequeue` from the `idleQueue` mapping that has no elements in its `Queue`, and thus giving us the `queue empty` thrown exception.

**Comment by tbrown1979** [comment]
Gonna see if we can figure this one out.

**Comment by calvinbrown085** [comment]
Should we close this issue?

**Comment by ChristopherDavenport** [comment]
Yes was addressed by #1922. Thank you!
